Pidgin.Examples/Expr/ExprParser.cs

```diff
@@ -0,0 +1,182 @@
+                    Operator.InfixL(Try(String("<>").Between(SkipWhitespaces)).Then(Return<Func<Expr,Expr,Expr>>((x,y)=>new Ne("<>",x,y)))),
+                    Operator.InfixL(String("!=").Between(SkipWhitespaces).Then(Return<Func<Expr,Expr,Expr>>((x,y)=>new Ne(x,y)))),
+                    Operator.InfixL(String(">").Between(SkipWhitespaces).Then(Return<Func<Expr,Expr,Expr>>((x,y)=>new Gt(x,y)))),
+                    Operator.InfixL(String("<").Between(SkipWhitespaces).Then(Return<Func<Expr,Expr,Expr>>((x,y)=>new Lt(x,y)))),
+                    Operator.InfixL(Try(String("==")).Or(String("=")).Between(SkipWhitespaces).Then(Return<Func<Expr,Expr,Expr>>((x,y)=>new Eq(x,y)))),
+                    Operator.InfixL(Try(String("in").Between(SkipWhitespaces).Before(Lookahead(LBrace))).Then(Return<Func<Expr,Expr,Expr>>((x,y)=>new In(x,y)))),
+                },
+                // logical. TODO: check how to separate arithmatic, comparison and logical ops
+                new[] {
+                    Operator.Prefix(Char('!').Between(SkipWhitespaces).Then(Return<Func<Expr,Expr>>(x=>new Not(x)))),
+                },
+                new[] {
+                    Operator.InfixL(String("&&").Between(SkipWhitespaces).Then(Return<Func<Expr,Expr,Expr>>((x,y)=>new And(x,y)))),
+                },
+                new[] {
+                    Operator.InfixL(String("||").Between(SkipWhitespaces).Then(Return<Func<Expr,Expr,Expr>>((x,y)=>new Or(x,y)))),
+                },
+                new[] {
+                    Operator.InfixR(If.Between(SkipWhitespaces).Then(Return<Func<Expr,Expr,Expr>>((x,y)=>
+                    {
+                        if (pyTerStack.Count==0) throw new InvalidOperationException("Unbalanced Python-style Ternary Operators, fewer 'else' parts than 'then' parts");
+                        pyTerStack.Pop();
+                        return TerUtils.FormPyTerExpr(x, y);
+                    }))),
+                    Operator.InfixR(Else.Between(SkipWhitespaces).Then(Return<Func<Expr,Expr,Expr>>((x,y)=>
+                    {
+                        pyTerStack.Push(true);
+                        return new PyTerExprElse(x, y);
+                    }))),
+                    Operator.InfixR(Try(Lookahead(Not(Char('?').Before(Char('?'))))).Then(Char('?').Between(SkipWhitespaces)).Then(Return<Func<Expr,Expr,Expr>>((x,y) =>
+                    {
+                        if (cTerStack.Count==0) throw new InvalidOperationException("Unbalanced C-style Ternary Operators, fewer 'else' parts than 'then' parts");
+                        cTerStack.Pop();
+                        return TerUtils.FormCTerExpr(x, y);
+                    }))),
+                    Operator.InfixR(Char(':').Between(SkipWhitespaces).Then(Return<Func<Expr,Expr,Expr>>((x,y)=>
+                    {
+                        cTerStack.Push(true);
+                        return new CTerElseExpr(x, y);
+                    }))),
+                },
```

Comment:
I'd use a unary postfix operator for `?:`. Something like

```csharp
// '?', then two expressions separated by a ':'
Operator.Postfix(
    Char('?')
        .Then(Rec(() => expr))
        .Then(Char(':'))
        .Then(Rec(() => expr))
)
```

This saves you from manually tracking state in a stack and reassociating the parse tree.

Reply:
I couldn't get this to compile.  The types were not auto deduced therefore I had to specify, some, but eventually this 
            exprParser = ExpressionParser.Build(termParser, new[]
            {
                Operator.Postfix<char,Expr>(Map<char,char,Expr,char,Expr,Expr>((_a,t,_b,f) => 
                    cond => (Expr)new TernaryIf(cond, t, f),  // <== error
                    Char('?'), Rec(()=> exprParser), Char(':'), Rec(()=>exprParser))),
            });

which fails to compile
CS1660 C# Cannot convert lambda expression to type because it is not a delegate type

Reply:
Shouldn’t that be `Map<char, Expr, char, Expr, Func<Expr, Expr>>(...)`?

Reply:
Or you could write down fewer types by explicitly casting the lambda to a `Func<Expr, Expr>`.